Хамидуллина Аниса

catsAndSuicide opened the pull request:
@lurlrlrl

catsAndSuicide committed:
1) Созданы интерфейс ICloudLayouter, классы CircularCloudLayouter и CircularCloudLayouterTests.

2) Теперь первый прямоугольник располагается так, что его центр совпадает с центром облака.

catsAndSuicide committed:
1) Добавлен новый тест, проверяющий, что прямоугольники располагаются в разных местах

2) В класс CircularCloudLayouter добавлен новый метод GetNextPoint, возвращающий друг за другом точки на окружности, которая меняется на окружность с большим радиусом при завершении обхода. Теперь прямоугольники располагаются в разных точках облака.

catsAndSuicide committed:
1) Создан класс Cloud, который с помощью класса CircularCloudLayouter умеет парсить файл с размерами прямоугольников в список прямоугольников.

2) Создан класс CloudVisualizer, который умеет рисовать экземпляр класса Cloud.
Оба класса - вспомогательные, чтобы была возможность оценить работу алгоритма расстановки.

catsAndSuicide committed:
Добавлена проверка на пересечение прямоугольников. Теперь они располагаются вдоль окружностей так, что не пересекаются.

catsAndSuicide committed:
Добавлен метод MoveToCenter. Теперь облако более плотное. Соответсвующая проверка добавлена.

catsAndSuicide committed:
Теперь при получении очередной точки окружности рассматриваем четыре возможных случая: это верхняя левая, верхняя правая, нижняя левая или нижняя правая вершина прямоугольника. Это позволяет делать облако более плотным.

catsAndSuicide committed:
1) Небольшой рефакторинг

2) Удален избыточный тест, добавлено рандомизированное добавление прямоугольников в раскладку

catsAndSuicide committed:
1) Исправлены ошибки в одном из тестов

2) Добавлена новая функциональность: теперь если тест не проходит, сохраняется текущая раскладка. Добавлен новый тест для демонстрации

lurlrlrl commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,47 @@
+                var sizeStr = reader.ReadLine().Split(' ');
+                sizes.Add(new Size(int.Parse(sizeStr[0]), int.Parse(sizeStr[1])));
+            }
+            reader.Close();
```
До вызова .Close() может выпасть иссключение

lurlrlrl commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,47 @@
+            while (!reader.EndOfStream)
+            {
+                var sizeStr = reader.ReadLine().Split(' ');
+                sizes.Add(new Size(int.Parse(sizeStr[0]), int.Parse(sizeStr[1])));
```
Можно использовать yield return

lurlrlrl commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace TagsCloudVisualization
+{
+    public class Cloud
```
Не понял смысл этого класса, вся логика у него расположена в конструкторе, а потом из него просто достаётся список прямоугольников. Я думаю, если заменить его на статическую функцию код будет выглядить проще.

lurlrlrl commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Здесь и в других местах усть неиспользуемые using

catsAndSuicide committed:
1) Появились классы RectangleExtensions и Utils, содержащие в себе часть кода, которая раньше была в CircularCloudLayouter. Это позволило сделать класс CircularCloudLayouter более закрытым и при этом использовать данный код сразу из CircularCloudLayouter и CircularCloudLayouterTests.

2) Теперь CloudVisualizer обращается к CircularCloudLayouter только через интерфейс ICloudLayouter.
3) Класс Cloud удален и переделан в статический метод в классе CloudVisualizer.
4) Исправлены другие мелкие замечания.

lurlrlrl commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        [Test]
+        public void PutNextRectangle_AsCloseAsPossible_IfRectangleDoNotIntersectOther()
```
Не понял, что означает в названии IfRectangleDoNotIntersectOther

catsAndSuicide replied:
Подразумевалось "IfRectangleDoesNotIntersectOther" - так близко, чтобы прямоугольник при этом не пересекал другие прямоугольники. Пыталась как-то придумать третью часть названия, чтобы она описывала условия, при которых должно выполняться проверяемое свойство. Можно ли вообще избавиться от этой части и назвать тест просто "PutNextRectangle_AsCloseAsPossible"?

lurlrlrl replied:
Да можно. Лично я предпочитаю названия которые близки к естественному языку, например я бы просто назвал Should_put_rectangles_as_close_as_possible

catsAndSuicide commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+
+        private Rectangle PutFirstRectangle(Size rectangleSize)
```
В обычном алгоритме размещения я для полученной точки перебираю четыре варианта: когда это верхняя левая вершина, верхняя правая и так далее - так повышается шанс, что я смогу в этой области разместить данный прямоугольник, а потом подвинуть поближе к центру. Тем временем первый прямоугольник нужно разместить, так чтобы его _центр_ совпадал с точкой.

lurlrlrl replied:
А после сдвига центр не совпадёт с точкой? Ок, можно оставить как есть

catsAndSuicide commented on cs/TagsCloudVisualization/RectangleExtensions.cs:
```diff
@@ -0,0 +1,48 @@
+            var rectCenter = new Point(
+                rectangle.X + rectangle.Width / 2,
+                rectangle.Y + rectangle.Height / 2);
+            return Math.Sqrt((point.X - rectangle.X) ^ 2 + (point.Y - rectCenter.Y) ^ 2);
```
Ой, да, это должно было быть возведение в степень.

lurlrlrl commented on cs/TagsCloudVisualization/CloudVisualizer.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            using (var image = new Bitmap(imageSize.Width, imageSize.Height))
+            {
+                var g = Graphics.FromImage(image);
```
Graphics тоже IDisposable

lurlrlrl commented on cs/TagsCloudVisualization/CloudVisualizer.cs:
```diff
@@ -0,0 +1,56 @@
+            {
+                var center = GetPointFromLine(reader.ReadLine());
+                var layouter = new CircularCloudLayouter(center);
+                var sizes = new List<Size>();
```
Переменная не используется

catsAndSuicide committed:
Внесла исправления согласно замечаниям. Добавила тест, проверяющий, что форма облака близка к кругу.

lurlrlrl commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,114 @@
+        public Rectangle PutNextRectangle(Size rectangleSize)
+        {
+            if (currentRectangles.Count == 0)
+                return PutFirstRectangle(rectangleSize);
```
В методе нет ни одной пустой строки, я бы добавил пару пустых строчек для логически разных частей метода. По-моему так бы легче читалось

lurlrlrl commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,114 @@
+            foreach (var direction in DirectionUtils.GetAllDirections())
+            {
+                var tempRectangle = rectangle.GetMovedCopy(direction, shift);
+                if (tempRectangle.IntersectsWithRectangles(currentRectangles)
```
По-моему если инвертировать if и написать:
```                var tempRectangle = rectangle.GetMovedCopy(direction, shift);
                if (!tempRectangle.IntersectsWithRectangles(currentRectangles) && !(tempRectangle.GetDistanceToPoint(Center) >= rectangle.GetDistanceToPoint(Center))) 
                    return tempRectangle;```
То будет лучше читаться

lurlrlrl commented on cs/TagsCloudVisualization/RectangleExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+{
+    public static class RectangleExtensions
+    {
+        public static bool IntersectsWithRectangles(this Rectangle currentRectangle, IEnumerable<Rectangle> rectangles)
```
У Rectangle уже есть метод IntersectsWith, который принимает Rectangle.
По-моему этот метод можно назвать также, в некоторых языках есть ограничения на перезагрузку методов и такое бы было сделать нельзя, но в C# я бы назвал так.
Это просто комментарий, а не призыв переписать

lurlrlrl commented on cs/TagsCloudVisualization/CloudVisualizer.cs:
```diff
@@ -0,0 +1,55 @@
+            using (var image = new Bitmap(imageSize.Width, imageSize.Height))
+            using (var g = Graphics.FromImage(image)) 
+            { 
+                var pen = new Pen(Color.Black);
```
А Pen это тоже IDisposable

lurlrlrl commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,135 @@
+
+        private int GetCloudHorizontalDiameter()
+        {
+            var top = rectangles.Select(rectangle => rectangle.Top).Min();
```
Можно использовать перезагрузку Min, которая принимает лямбду

lurlrlrl commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        [Test]
+        public void TestOnlyForDemonstrationThirdTask()
```
Публичные методы должны быть выше приватных

catsAndSuicide committed:
Рефакторинг
